src/fastpath/fastpath.c

```diff
@@ -902,3 +902,252 @@ void NORETURN fastpath_vm_fault(vm_fault_type_t type)
+    asid_t asid = cap_vspace_cap_get_capVSMappedASID(newVTable);
+    asid_map_t asid_map = findMapForASID(asid);
+    if (unlikely(asid_map_get_type(asid_map) != asid_map_asid_map_vspace ||
+                 VSPACE_PTR(asid_map_asid_map_vspace_get_vspace_root(asid_map)) != cap_pd)) {
+        slowpath_irq();
+    }
+#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
+    /* Ensure the vmid is valid. */
+    if (unlikely(!asid_map_asid_map_vspace_get_stored_vmid_valid(asid_map))) {
+        slowpath_irq();
+    }
+    /* vmids are the tags used instead of hw_asids in hyp mode */
+    stored_hw_asid.words[0] = asid_map_asid_map_vspace_get_stored_hw_vmid(asid_map);
+#else
+    stored_hw_asid.words[0] = asid;
+#endif
+#endif
+
+#ifdef CONFIG_ARCH_RISCV
+    /* Get HW ASID */
+    stored_hw_asid.words[0] = cap_page_table_cap_get_capPTMappedASID(newVTable);
+#endif
+
+    /*  Point of no return */
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+    ksKernelEntry.is_fastpath = true;
+#endif
+
+
+    if (idle) {
+        /* Cancel the IPC that the signalled thread is waiting on */
+        cancelIPC_fp(dest);
+    } else {
+        /* Dequeue dest from the notification queue */
+        ntfn_queue_dequeue_fp(dest, ntfn_ptr);
+    }
+
+    /* Wake up the signalled thread and tranfer badge */
+    setRegister(dest, badgeRegister, badge);
+    thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+
+    /* Donate SC if necessary. The checks for this were already done before
+     * the point of no return */
+    maybeDonateSchedContext_fp(dest, sc);
+
+    /* Left this in the same form as the slowpath. Not sure if optimal */
+    if (sc_sporadic(dest->tcbSchedContext)) {
+        assert(dest->tcbSchedContext != NODE_STATE(ksCurSC));
+        if (dest->tcbSchedContext != NODE_STATE(ksCurSC)) {
+            refill_unblock_check(dest->tcbSchedContext);
+        }
+    }
+
+    if (!schedulable) {
+        mask_ack_bail(irq);
+        UNREACHABLE();
+    }
+
+    if (ksCurDomain != dest->tcbDomain SMP_COND_STATEMENT( || sc->scCore != getCurrentCPUIndex())) {
+        crossnode = true;
+    }
+
+    if (dest->tcbPriority > NODE_STATE(ksCurThread)->tcbPriority || crossnode) {
+        /* we're going to switch threads - charge the current thread */
+        updateTimestamp();
+        ticks_t prev = getNextInterrupt();
+        /* charge the previous thread */
+        if (checkBudget()) {
+            commitTime();
+        }
+
+        if (isSchedulable(NODE_STATE(ksCurThread))) {
+            SCHED_ENQUEUE_CURRENT_TCB;
+        }
+        switchToThread_fp(dest, cap_pd, stored_hw_asid);
+        /* only set the next irq if we really have to */
+        ticks_t next = getNextInterrupt();
+        if (next < prev) {
+            setDeadline(next - getTimerPrecision());
+        }
+        /* update sc */
+        NODE_STATE(ksCurSC) = NODE_STATE(ksCurThread)->tcbSchedContext;
+    } else {
+        SCHED_ENQUEUE(dest);
+    }
```

Comment:
crossnode and prio(curr) > (dest) result in an enqueue while prio(curr) == prio(dest) results in an append.